pallets/api/src/fungibles/tests.rs

```diff
@@ -342,3 +342,98 @@ fn set_metadata_token(
 ) {
 	assert_ok!(Assets::set_metadata(signed(owner), token, name, symbol, decimals));
 }
+
+mod read_weights {
+	use frame_support::weights::Weight;
+
+	use super::*;
+	use crate::fungibles::{weights::WeightInfo, Config};
+
+	struct ReadWeightInfo {
+		allowance: Weight,
+		balance_of: Weight,
+		token_decimals: Weight,
+		token_name: Weight,
+		token_symbol: Weight,
+		total_supply: Weight,
+		token_exists: Weight,
+	}
+
+	impl ReadWeightInfo {
+		fn new() -> Self {
+			Self {
+				allowance: Fungibles::weight(&Allowance {
+					token: TOKEN,
+					owner: ALICE,
+					spender: BOB,
+				}),
+				balance_of: Fungibles::weight(&BalanceOf { token: TOKEN, owner: ALICE }),
+				token_decimals: Fungibles::weight(&TokenDecimals(TOKEN)),
+				token_name: Fungibles::weight(&TokenName(TOKEN)),
+				token_symbol: Fungibles::weight(&TokenSymbol(TOKEN)),
+				total_supply: Fungibles::weight(&TotalSupply(TOKEN)),
+				token_exists: Fungibles::weight(&TokenExists(TOKEN)),
+			}
+		}
+	}
+
+	#[test]
+	fn ensure_read_matches_benchmarks() {
+		let ReadWeightInfo {
+			allowance,
+			balance_of,
+			token_decimals,
+			token_name,
+			token_symbol,
+			total_supply,
+			token_exists,
+		} = ReadWeightInfo::new();
+
+		assert_eq!(allowance, <Test as Config>::WeightInfo::allowance());
+		assert_eq!(balance_of, <Test as Config>::WeightInfo::balance_of());
+		assert_eq!(token_decimals, <Test as Config>::WeightInfo::token_decimals());
+		assert_eq!(token_name, <Test as Config>::WeightInfo::token_name());
+		assert_eq!(token_symbol, <Test as Config>::WeightInfo::token_symbol());
+		assert_eq!(total_supply, <Test as Config>::WeightInfo::total_supply());
+		assert_eq!(token_exists, <Test as Config>::WeightInfo::token_exists());
+	}
+
+	// These types read from the `AssetMetadata` storage.
+	#[test]
+	fn ensure_asset_metadata_variants_match() {
+		let ReadWeightInfo { token_decimals, token_name, token_symbol, .. } = ReadWeightInfo::new();
+
+		assert_eq!(token_decimals, token_name);
+		assert_eq!(token_decimals, token_symbol);
+	}
+
+	// These types read from the `Assets` storage.
+	#[test]
+	fn ensure_asset_variants_match() {
+		let ReadWeightInfo { total_supply, token_exists, .. } = ReadWeightInfo::new();
+
+		assert_eq!(total_supply, token_exists);
+	}
+
+	// Proof size is based on `MaxEncodedLen`, not hardware.
+	#[test]
+	fn ensure_expected_proof_size_does_not_change() {
```

Comment:
If you want to double check  this, please try on your machine!
```
./target/release/pop-node \
benchmark \
pallet \
--chain=dev \
--wasm-execution=compiled \
--pallet=fungibles \
--steps=50 \
--repeat=20 \
--json \
--template \
./scripts/pallet-weights-template.hbs \
--output=./pallets/api/src/fungibles/weights.rs \
--extrinsic=
```

Reply:
Would be nice if we also assert to the maxencodedlen of the structs from pallet assets maybe? How you did it previously

Reply:
please see latest commit

Reply:
Yeah. Tbh when I saw your comment last night I thought the same, so had the same misunderstanding.

My view is that if we rely on the accuracy of benchmarks/weights for dispatchables, recreating it for reads seems unnecessary. There is a framework there for benchmarking and getting in-depth results out of it, so its reasonable to rely on it as that is what underpins the ecosystem. Fine for some additional sanity checks tho.

Reply:
Whilst I admire the level of detail, clarity and accuracy of these tests, my only reservation would be whether this would become an expectation moving forward. My concern would be the additional time required to implement for all the other APIs to remain consistent versus the value it brings.

Reply:
Same, therefore I think it is best to remove. Thank for the clear detailed docs and tests though Peter, was able to immediately grasp the detailed workings of it!

Reply:
Thanks for the clear detailed docs and tests though Peter, was able to immediately grasp the detailed workings of it!

Reply:
It was quite tedious to do! Although thankful for the learnings :) The last test that verifies the proof size remains the same will effectively provide the same testing outcome (make sure the max_encoded_len does not change) please see the latest commit, reverting the changes.